plugins/soundsourcemediafoundation/audiosourcemediafoundation.cpp

```diff
@@ -0,0 +1,558 @@
+ */
+inline qint64 mfFromSeconds(qreal sec) {
+    return sec * 1e7;
+}
+
+/**
+ * Convert a 100ns Media Foundation value to a frame offset.
+ */
+inline qint64 frameFromMF(qint64 mf) {
+    return static_cast<qreal>(mf) * kSampleRate / 1e7;
+}
+
+/**
+ * Convert a frame offset to a 100ns Media Foundation value.
+ */
+inline qint64 mfFromFrame(qint64 frame) {
+    return static_cast<qreal>(frame) / kSampleRate * 1e7;
+}
+
+/** Microsoft examples use this snippet often. */
+template<class T> static void safeRelease(T **ppT) {
+    if (*ppT) {
+        (*ppT)->Release();
+        *ppT = NULL;
+    }
+}
+
+}
+
+AudioSourceMediaFoundation::AudioSourceMediaFoundation()
+        : m_hrCoInitialize(E_FAIL), m_hrMFStartup(E_FAIL),
+                m_pReader(NULL), m_pAudioType(NULL), m_wcFilename(
+                NULL), m_nextFrame(0), m_leftoverBuffer(NULL), m_leftoverBufferSize(
+                0), m_leftoverBufferLength(0), m_leftoverBufferPosition(0), m_mfDuration(
+                0), m_iCurrentPosition(0), m_dead(false), m_seeking(false) {
+
+    // these are always the same, might as well just stick them here
+    // -bkgood
+    // AudioSource properties
+    setChannelCount(kNumChannels);
+    setFrameRate(kSampleRate);
+
+    // presentation attribute MF_PD_AUDIO_ENCODING_BITRATE only exists for
+    // presentation descriptors, one of which MFSourceReader is not.
+    // Therefore, we calculate it ourselves, assuming 16 bits per sample
+    setBitrate(frames2samples(kBitsPerSampleForBitrate * kSampleRate) / 1000);
+}
+
+AudioSourceMediaFoundation::~AudioSourceMediaFoundation() {
+    close();
+}
+
+AudioSourcePointer AudioSourceMediaFoundation::open(QString fileName) {
+    AudioSourceMediaFoundation* pAudioSourceMediaFoundation(new AudioSourceMediaFoundation);
+    AudioSourcePointer pAudioSource(pAudioSourceMediaFoundation); // take ownership
+    if (OK == pAudioSourceMediaFoundation->postConstruct(fileName)) {
+        // success
+        return pAudioSource;
+    } else {
+        // failure
+        return AudioSourcePointer();
+    }
+}
+
+Result AudioSourceMediaFoundation::postConstruct(QString fileName) {
+    if (sDebug) {
+        qDebug() << "open()" << fileName;
+    }
+
+    // Initialize the COM library.
+    m_hrCoInitialize = CoInitializeEx(NULL,
+            COINIT_APARTMENTTHREADED | COINIT_DISABLE_OLE1DDE);
+    if (FAILED(m_hrCoInitialize)) {
+        qWarning() << "SSMF: failed to initialize COM";
+        return ERR;
+    }
+    // Initialize the Media Foundation platform.
+    m_hrMFStartup = MFStartup(MF_VERSION);
+    if (FAILED(m_hrCoInitialize)) {
+        qWarning() << "SSMF: failed to initialize Media Foundation";
+        return ERR;
+    }
+
+    QString qurlStr(fileName);
+    // http://social.msdn.microsoft.com/Forums/en/netfxbcl/thread/35c6a451-3507-40c8-9d1c-8d4edde7c0cc
+    // gives maximum path + file length as 248 + 260, using that -bkgood
+    m_wcFilename = new wchar_t[248 + 260];
```

Comment:
It looks like we can make m_wcFilename local. 
Or we can ditch all this cnversion stuff and use 

MFCreateSourceReaderFromURL(fileName.toStdWString(), NULL, &m_pReader);